Events subsystem

Abyss777 opened the pull request:
What have done:

A model and a database store (+ added liquibase changelog)
A base class for events handlers
Three example handlers in a separate package org.traccar.events

`CommandResultEventHandler` - just fire an event if there is "result" in position attributes

`OverspeedEventHandler` - fire an event if the device exceeds the speed (has draft for by group and by device speedlimiting, needs model and UI extension)

`MotionEventHandler` - fire an event on device motion, used new device field and extended updateDeviceStatus

Last two can suppres repeated events.

`ConnectionManager` fire events on offline/online, sends new events in websocket.

API represents events by ID and last events for period (in seconds).

In the last commit I've added synchronization `devicesById` and database, I think my way is cheaper than use `updateDeviceCache`. It is needed for correct motion detection and in other situations.

tananaev commented on database/changelog-3.6.xml:
```diff
@@ -0,0 +1,34 @@
+      <column name="type" type="VARCHAR(128)">
+        <constraints nullable="false" />
+      </column>
+      <column name="eventtime" type="TIMESTAMP">
```
I think we can have `serverTime` and `deviceTime` similar to position model.

Abyss777 replied:
Is it really need? We can pool `serverTime` and `deviceTime` from the original position. And we don't have device time for online/offline events.

tananaev replied:
I think it's just less obvious name. Is event time device time or is it server time? Name doesn't say anything.

tananaev commented on setup/unix/traccar.xml:
```diff
@@ -17,6 +17,13 @@
     <entry key='logger.level'>all</entry>
     <entry key='logger.file'>/opt/traccar/logs/tracker-server.log</entry>
 
+    <entry key='event.suppressrepeated'>60</entry>
```
use lower camel case for parameter names

tananaev commented on setup/unix/traccar.xml:
```diff
@@ -165,6 +172,23 @@
+    </entry>
+
+    <entry key='database.selectLastEvents'>
+        SELECT * FROM events WHERE deviceId = :deviceId AND type LIKE :type AND eventTime >= NOW() - INTERVAL :interval SECOND ORDER BY eventTime DESC;
```
Is INTERVAL universal for all SQL databases?

tananaev commented on src/org/traccar/api/resource/EventResource.java:
```diff
@@ -0,0 +1,39 @@
+            @QueryParam("interval") long interval) throws SQLException {
+        Context.getPermissionsManager().checkDevice(getUserId(), deviceId);
+        return Context.getDataManager().getLastEvents(deviceId, type, interval);
+        }
```
indentation problem

Abyss777 replied:
The function params is after 8 spaces from `public`, the code is after 4 spaces from `public`. In the other resources is the same. I have to move third param to new line because of checkstyle (string long then 120 symbols). Or are you talking about something else?

tananaev replied:
I'm talking about something else. I'm talking about closing bracket of the method body.

tananaev commented on src/org/traccar/events/CommandResultEventHandler.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    protected Event analizePosition(Position position) {
+        Object cmdResult = position.getAttributes().get(Position.KEY_RESULT);
```
rename cmdResult to commandResult

tananaev commented on src/org/traccar/events/OverspeedEventHandler.java:
```diff
@@ -0,0 +1,79 @@
+import org.traccar.model.Position;
+import org.traccar.model.Server;
+import org.traccar.helper.Log;
+//import org.traccar.model.Device;
```
remove comments if imports are not needed

tananaev commented on src/org/traccar/events/OverspeedEventHandler.java:
```diff
@@ -0,0 +1,79 @@
+            String speedUnit = server.getSpeedUnit();
+            if (speedUnit != null) {
+                switch (speedUnit) {
+                    case "kmh" : globalSpeedLimit = UnitsConverter.knotsFromKph(globalSpeedLimit);
```
config shouldn't depend on server settings. Just pick one unit (probably km/h makes more sense)

tananaev commented on src/org/traccar/events/OverspeedEventHandler.java:
```diff
@@ -0,0 +1,79 @@
+        boolean valid = position.getValid();
+        double deviceSpeedLimit = 0;
+        double groupSpeedLimit = 0;
+//        Device device = Context.getIdentityManager().getDeviceById(position.getDeviceId());
```
remove commented code

tananaev commented on src/org/traccar/model/Event.java:
```diff
@@ -0,0 +1,74 @@
+        this.id = id;
+    }
+
+    public static final String COMMAND_RESULT = "command-result";
```
add TYPE_ prefix

tananaev commented:
Added comments. Have you run `mvn site` to check the code?

Abyss777 commented:
> Have you run mvn site to check the code?
Yes, sure. I'm doing it before every commit to master.

I've removed request used "INTERVAL" because it is really not universal. And it is easier do the same in code than find cross-database analog.

I've changed TYPE_* strings to lower camel case for uniformity.

And done other proposal except two commented.

Abyss777 commented:
Oh, how can I miss the bracket.
Renamed eventTime to serverTime.

tananaev commented:
Merged, thanks.
